### PR TITLE
HPCC-13828 Generate key comparison functions for sorts

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday gavin.halliday@lexisnexis.com

### CODE REVIEW

**Comment by ghalliday** [comment]
@jakesmith this implements the new functions.  I haven't done sliding join yet (but it also disables serialize in that case).

File: ecl/hqlcpp/hqlhtcpp.cpp [hunk 1]
```diff
@@ -11616,97 +11616,124 @@ void HqlCppTranslator::generateSerializeFunction(BuildCtx & ctx, const char * fu
...
+}
 
-            BuildCtx classctx(nestedctx);
-            beginNestedClass(classctx, memberName, "ISortKeySerializer");
+void HqlCppTranslator::generateSerializeKey(BuildCtx & nestedctx, node_operator side, SerializeKeyInfo & keyInfo, const DatasetReference & dataset, bool generateCompares)
+{
+    //check if there are any ifblocks, and if so don't allow it.  Even more accurate would be no join fields used in ifblocks
+    //IHqlExpression * record = dataset.queryDataset()->queryRecord();
+    const char * sideText = (side == no_none) ? "" : (side == no_left) ? "Left" : "Right";
+    StringBuffer s;
```
Comment (jakesmith) [inline ecl/hqlcpp/hqlhtcpp.cpp]: could be closer to use (on line 11718) and no need to .clear() call.

**Comment by jakesmith** [comment]
Looks good, I can't see anything significant.
I have regression tested this with my changes using these new helper methods and everything now passes.
I will release my PR based on your branch in a moment.
@ghalliday

**Comment by ghalliday** [comment]
@richardkchapman for you to review.

File: ecl/hqlcpp/hqlcset.cpp [hunk 2]
```diff
@@ -835,6 +835,13 @@ IHqlExpression * InlineLinkedDictionaryCursor::getFirstSearchValue(IHqlExpressio
     return matched;
 }
 
+BoundRow * InlineLinkedDictionaryCursor::buildIterateLoop(BuildCtx & ctx, bool needToBreak)
```
Comment (richardkchapman) [inline ecl/hqlcpp/hqlcset.cpp]: Is this related?

Comment (ghalliday) [inline ecl/hqlcpp/hqlcset.cpp]: Yes - when generating the compare function for something containing a dictionary it was generating an internal error.

File: ecl/hqlcpp/hqlhtcpp.cpp (hunk 1 above)
Comment (ghalliday) [inline ecl/hqlcpp/hqlhtcpp.cpp]: The use of clear() is there so that all calls are consistent.
It could be closer to where it is needed, I'm not sure it is worth changing, but could do.
